refactor: Replace redesign confirmation BottomModal with BottomSheet

digiwand opened the pull request:
<!--
Please submit this PR as a draft initially.
Do not mark it as "Ready for review" until the template has been completely filled out, and PR status checks have passed at least once.
-->


## **Description**

Updates:
- Replaces BottomModal with BottomSheet in redesign Confirm page
  - Remove min height to prevent redesign Confirm page from showing excess space
  - Allow BottomSheet styles (rounded corners, footer bottom styles, and device padding detection padding) rather than custom BottomModal styles
 - Update BlockaidBanner to remove surrounding margin to allow for better reusability
   - Replace BlockaidBanner instances margin overrides with surrounding container margins
 - Update BottomSheet to pass style to BottomSheetDialog sheet cc: @brianacnguyen 
   - Note: I think we don't need the additional background colors for header and footer since the parent, BottomSheet animated view sets the background color. Once I removed them, it triggered many snapshot tests and code owner reviews from 5 additional teams. Instead of making this change, we will continue to update the bottomsheet, footer, and header when applicable

Fixes:
 - Fix TransactionReview BlockaidBanner negative padding → turn to positive padding to allow space between next element
 - Fix Confirm.test console.errors
 
## **Related issues**

Fixes: https://github.com/MetaMask/metamask-mobile/issues/13267
Follow-up Related: https://github.com/MetaMask/metamask-mobile/pull/13783 (Fix testID)
Follow-up Related: https://github.com/MetaMask/metamask-mobile/pull/13913 (Scroll and close by backdrop click or drag down)

Follow-up - Related Issue: https://github.com/MetaMask/metamask-mobile/issues/12656

Cherry-pick PRs:
1/3: https://github.com/MetaMask/metamask-mobile/pull/13914 (BottomModal → BottomSheet)
2/3: https://github.com/MetaMask/metamask-mobile/pull/13926 (bitrise failure testID)
3/3: https://github.com/MetaMask/metamask-mobile/pull/13927 (scroll and close on swipe down & backdrop click)

## **Manual testing steps**

Test BottomSheets and BlockaidBanner alerts work as expected. Test redesign confirmation page works as expected

## **Screenshots/Recordings**

<!-- If applicable, add screenshots and/or recordings to visualize the before and after of your change. -->

### Before using BottomModal and no scroll behavior

https://github.com/user-attachments/assets/0d4c5453-5824-47fd-8aa5-7405cf19de53

### After using BottomSheet with scroll behavior 

<img width="320" src="https://github.com/user-attachments/assets/a908f81f-1a28-4e8c-966a-a390430dc7d9">

https://github.com/user-attachments/assets/8f8e2344-7706-49a4-b773-a5226b7826d4

### Before 70% min height

<img width="320" src="https://github.com/user-attachments/assets/94509d26-5e20-414d-942f-56938620f90a">

### After remove  70% min height 

<img width="320" src="https://github.com/user-attachments/assets/641022c5-6f23-47a5-9f00-4e9358a96d3b">

### **Before Transaction Review BlockaidBanner negative padding**

<img width="320" src="https://github.com/user-attachments/assets/c057c4e7-c7a7-42f2-8251-b4f2c075a802">

### **After Transaction Review BlockaidBanner positive padding **

<img width="320" src="https://github.com/user-attachments/assets/de1ecadb-c7ef-49e1-8ae2-6ce1b5d16997">

### Before Confirm.test console errors

![CleanShot 2025-02-09 at 01 35 46@2x](https://github.com/user-attachments/assets/2653a073-b0de-4c2d-bcea-dfbc11d38246)
![CleanShot 2025-02-09 at 01 35 54@2x](https://github.com/user-attachments/assets/f9ce410c-42c5-4738-b99f-f871456eb82f)

### After Confirm.test no console.errors

![CleanShot 2025-02-09 at 01 49 47@2x](https://github.com/user-attachments/assets/d3bf6580-553d-4b14-96ab-3f529d17db67)


## **Pre-merge author checklist**

- [ ] I’ve followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Mobile Coding Standards](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [ ] I've completed the PR template to the best of my ability
- [ ] I’ve included tests if applicable
- [ ] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [ ] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

digiwand committed:
refactor: rm background color form bottom sheet footer and header
no need because BottomSheetDialog provides background color

digiwand committed:
refactor: use BottomSheet for redesign Confirm

digiwand committed:
refactor: replace BlockaidBanner style prop uses
- affects only old confirmations
- fix marginBottom on TransactionReview

digiwand committed:
test: update snapshots following BottomSheet updates

digiwand committed:
test:fix: redesign Confirm mock SafeAreaProvider

digiwand committed:
Revert "refactor: rm background color form bottom sheet footer and header"

This reverts commit 04a974b13911c02a14142197b995ac6adde0f23b.

digiwand committed:
refactor: BottomSheet replace styleAnimatedView -> isBackgroundAlternative

digiwand committed:
test:fix: restore BottomSheet header and footer background

digiwand commented on app/components/Views/confirmations/Approval/__snapshots__/index.test.tsx.snap:
```diff
@@ -595,7 +595,7 @@ exports[`Approval render matches snapshot 1`] = `
                                                     style={
                                                       {
-                                                        "marginBottom": -8,
+                                                        "marginBottom": 8,
```
fixes missing padding between blockaid banner alert and account header

jpuri commented on app/components/Views/confirmations/components/SignatureRequest/index.js:
```diff
@@ -361,11 +361,12 @@ class SignatureRequest extends PureComponent {
+                securityAlertResponse={securityAlertResponse}
+                onContactUsClicked={this.onContactUsClicked}
+              />
+            </View>
```
I would suggest avoiding changing old pages

digiwand replied:
needed to make the change so that the reusable component does not apply it's own padding/margin. I think it's better to do this now than to continue this pattern of overriding the margin/padding through props

jpuri commented:
A small thing I see missing in screenshots this small rectangle at top of bottom modal, its required by the designs:

![Uploading Screenshot 2025-01-31 at 4.12.25 PM.png…]()

jpuri commented on app/components/Views/confirmations/Confirm/Confirm.styles.ts:
```diff
@@ -1,23 +1,9 @@
+    scrollView: {
       paddingHorizontal: 16,
-      paddingVertical: 24,
-      minHeight: '70%',
```
I am not if removing minHeight is good idea, let's check once with @SayaGT

digiwand replied:
### Before 70% min height

<img width="320" src="https://github.com/user-attachments/assets/94509d26-5e20-414d-942f-56938620f90a">

### After remove  70% min height 

<img width="320" src="https://github.com/user-attachments/assets/641022c5-6f23-47a5-9f00-4e9358a96d3b">

digiwand replied:
Confirmed we do not want fixed or min height

jpuri commented on app/components/Views/confirmations/components/TransactionReview/index.js:
```diff
@@ -113,9 +113,9 @@ const createStyles = (colors) =>
+      marginBottom: 8,
     },
   });
 
```
Same, I would suggest avoiding any change to old pages.

digiwand replied:
this minor change fixes this

### **Before Transaction Review BlockaidBanner negative padding**

<img width="320" src="https://github.com/user-attachments/assets/c057c4e7-c7a7-42f2-8251-b4f2c075a802">

### **After Transaction Review BlockaidBanner positive padding **

<img width="320" src="https://github.com/user-attachments/assets/de1ecadb-c7ef-49e1-8ae2-6ce1b5d16997">

jpuri commented on app/components/Views/confirmations/Confirm/Confirm.tsx:
```diff
@@ -22,18 +22,20 @@ const Confirm = () => {
-    <BottomModal canCloseOnBackdropClick={false}>
-      <View style={styles.container} testID={approvalRequest?.type}>
-        <View>
-          <Title />
```
We now have a mix of use of `BottomSheet` and `BottomModal` in  confirmation pages. Should we replace all occurence of `BottomModal` like [these](https://github.com/MetaMask/metamask-mobile/blob/10a519161471f06e8f954ea1529c33dfbbf45aa9/app/components/Views/confirmations/components/Confirm/Info/TypedSignV3V4/Simulation/components/NativeValueDisplay/NativeValueDisplay.tsx#L121).

digiwand replied:
That's correct.

That was the original ticket found in the description
```
Todo follow-up - Related Issue: https://github.com/MetaMask/metamask-mobile/issues/12656
```

I created a separate issue ticket after working on this initial transition. Note: It is not possible to put a BottomSheet over a BottomModal too which is why this one needs to be done first

jpuri reviewed (changes requested):
Changes look good, I added some feedback that need to be addressed.

digiwand commented on app/components/Views/confirmations/Confirm/Confirm.tsx:
```diff
@@ -22,18 +22,20 @@ const Confirm = () => {
-        <View>
-          <Title />
+    <BottomSheet
+      isInteractable={false}
```
re: grey rounded, rectangular UI at top of sheet
https://github.com/MetaMask/metamask-mobile/pull/13268#issuecomment-2626898070

If isInteractable is here true, then we show the gray rectangle. Dragging the UI down would close the UI. However, it doesn't seem we want the user to close the confirmation on drag down. 

<img width="320" src="https://github.com/user-attachments/assets/df51bac0-becc-47f6-84dd-2da76188a213">
 
Interestingly, the BottomModal supports this behavior where it is draggable, but not closable:

https://github.com/user-attachments/assets/537bc82a-0b06-45b4-9e71-150a26044fc3

This is not a behavior with our DS BotttomSheet though. What is the expected behavior for the Confirm page here? Should we use BottomSheet isInteractable false as provided by DS? cc: @SayaGT @MetaMask/design-system

digiwand replied:
discussed w/ DS and the team that we do want to continue using BottomSheet with isInteractable === false. no rectangle will be shown in this case

digiwand committed:
Merge branch 'main' into refactor-signature-use-bottom-sheet

digiwand committed:
fix: merge conflicts

digiwand committed:
test: rm async and update phrases Confirm.test

digiwand committed:
test: rm unused getAddressAccountType mock Confirm.test

digiwand committed:
test: rm unnecessary beforeEach clearAllMocks

digiwand committed:
test: Confirm.test add more mocks to remove console.errors

digiwand committed:
test: Confirm.test rm missing act warning

digiwand committed:
Merge branch 'main' into refactor-signature-use-bottom-sheet

digiwand committed:
refactor: revert isBackgroundAlternative ->  stylesDialogSheet Confirm BottomSheet

digiwand committed:
test:fix: bypass lint empty function

brianacnguyen reviewed (changes requested):
Request changes RE BottomSheet

digiwand commented on app/component-library/components/BottomSheets/BottomSheet/BottomSheet.tsx:
```diff
@@ -36,6 +36,7 @@ const BottomSheet = forwardRef<BottomSheetRef, BottomSheetProps>(
       children,
       onClose,
       onOpen,
+      isBackgroundAlternative = false,
```
hey @brianacnguyen, thank you for the review and reminder to push my updates! Apologies I didn't get to add this proposal before you reviewed. What do you think if we used stylesDialogSheet instead? This is because the style isn't for the general bottom sheet styles, but it's a specific style in BottomSheetDialog → sheet

digiwand replied:
Thanks @brianacnguyen! Lgtm! :)

digiwand committed:
fix: Confirm merge conflicts padding and height

digiwand committed:
fix: ConfirmFooter backgroundColor

digiwand committed:
Merge branch 'main' into refactor-signature-use-bottom-sheet

digiwand committed:
fix: confirm Footer import

digiwand committed:
Merge branch 'main' into refactor-signature-use-bottom-sheet

digiwand committed:
Merge branch 'main' into refactor-signature-use-bottom-sheet

digiwand committed:
fix: merge conflicts rm Confirm scrollView styles

digiwand committed:
fix: confirm scroll centering

digiwand committed:
fix: confirm scroll icon color

digiwand committed:
fix: confirm scroll performance and log warnings

digiwand committed:
fix: snapshots after removal of View + margin

digiwand committed:
Merge branch 'main' into refactor-signature-use-bottom-sheet

digiwand committed:
fix merge

digiwand committed:
Merge branch 'main' into refactor-signature-use-bottom-sheet

digiwand committed:
merge commit fixes

brianacnguyen committed:
Updated BottomSheetDialog style

digiwand committed:
Merge branch 'main' into refactor-signature-use-bottom-sheet

digiwand committed:
style: fix indent

digiwand committed:
fix: confirm footer styles
adding styles overwrites alignment applied styles (flex: 1)
also allow inherit disabled styles to be used

digiwand committed:
fix: Footer import LedgerContext

digiwand committed:
fix: bottom to top animation
removing this option seems to allow background to fade in naturally
if this doesn't work try animation: 'fade'

digiwand committed:
Merge branch 'main' into refactor-signature-use-bottom-sheet

brianacnguyen committed:
Removed unnecessary changes

OGPoyraz commented on app/components/Views/confirmations/Confirm/Confirm.styles.ts:
```diff
@@ -18,21 +21,6 @@ const styleSheet = (params: { theme: Theme }) => {
       justifyContent: 'space-between',
       paddingHorizontal: 16,
     },
-    modalContainer: {
```
I think we should keep `modalContainer` styles, remove everything except `paddingHorizontal:16` and keep the `ConfirmWrapped` wrapped with a `View`. 

This is current signature confirmation - needs horizontal padding.
<img width="515" alt="Screenshot 2025-02-26 at 20 42 13" src="https://github.com/user-attachments/assets/b7b615e7-b33a-4bac-bbb1-f06b6941b675" />

And this is staking 
<img width="559" alt="Screenshot 2025-02-26 at 20 49 00" src="https://github.com/user-attachments/assets/974eaa10-efab-49f6-a1e7-913ee3d66ff5" />

digiwand replied:
ty @OGPoyraz! updated the padding https://github.com/user-attachments/assets/25303f1d-f885-4b4f-8dd6-d4817c182629

how did you repro your staking confirmation? This is mine:

<img width="320" src="https://github.com/user-attachments/assets/25303f1d-f885-4b4f-8dd6-d4817c182629">

OGPoyraz replied:
@digiwand the one you see is the current staking confirmation.
In order to see the new one you need to set `export FEATURE_FLAG_REDESIGNED_STAKING_TRANSACTIONS='true'` in your `.js.env` file.

After last changes - staking deposit now looks more narrowed down
<img width="559" alt="Screenshot 2025-02-28 at 09 55 33" src="https://github.com/user-attachments/assets/56622129-6440-4b7b-b00a-e78a338e3800" />

However now the only change we need is to remove paddingHorizontal from `flatContainer`
https://github.com/MetaMask/metamask-mobile/blob/refactor-signature-use-bottom-sheet/app/components/Views/confirmations/Confirm/Confirm.styles.ts#L22

And then we are good to go. Tested the change above for you 

<img width="515" alt="Screenshot 2025-02-28 at 09 59 08" src="https://github.com/user-attachments/assets/1e8fd1c2-a683-4274-a6e0-beb775da900f" />

digiwand replied:
updated! Thanks @OGPoyraz for the info and update to support staking
https://github.com/MetaMask/metamask-mobile/pull/13268/commits/a7c539a37e5f3dd163c6480df9784bb292584c24

OGPoyraz commented on app/components/Views/confirmations/components/Confirm/Footer/Footer.styles.ts:
```diff
@@ -1,34 +1,15 @@
-      flexDirection: 'row',
+    base: {
+      backgroundColor: colors.background.alternative,
+      paddingHorizontal: 16,
```
I think `paddingHorizontal: 16` should be removed here since we apply horizontal padding only in containers, unless we really need them.

digiwand replied:
the reason this padding is here is because I moved the padding into the scrollable section. This allows the scrollbar to be further on the right than in the middle of the UI. I think this looks cleaner than the scrollbar with padding 16 from the right

<img width="320" src="https://github.com/user-attachments/assets/321e2fda-7fbe-4155-af1e-bdcd304b73e9">

OGPoyraz replied:
Thanks for clarification, that make sense

OGPoyraz commented on app/components/Views/confirmations/Confirm/Confirm.tsx:
```diff
@@ -6,30 +6,31 @@ import {
     <LedgerContextProvider>
       <Title />
       <ScrollView style={styles.scrollable}>
-        <TouchableWithoutFeedback>
+        <TouchableWithoutFeedback testID={testID}>
           <View style={styles.scrollableSection}>
```
Should we remove assigned styles from `scrollable` and `scrollableSection`?

digiwand replied:
thank you! I think this was a merge conflict. removed these two classes and re-added class to adding padding horizontal

jpuri commented on app/components/Views/confirmations/Confirm/Confirm.tsx:
```diff
@@ -62,10 +62,12 @@ export const Confirm = () => {
+      testID="modal-confirmation-container"
+    >
+      <ConfirmWrapped testID={approvalRequest?.type} styles={styles} />
+    </BottomSheet>
```
BottomModal has this behaviour of rejecting confirmation is closed `onClose={onReject}` can we ensure that in BottomSheet also, it will be nice to have BottomSheet interactable I think.

digiwand replied:
okay, this sounds good to me to add it back to ensure approval request is rejected on close. I'm thinking the only ways to close the modal are with the confirm/reject button or closing the app

digiwand committed:
Merge branch 'main' into refactor-signature-use-bottom-sheet

digiwand committed:
fix: reintroduce onClose onReject Confirm just in case

digiwand committed:
fix: missing scroll paddingHorizontal

digiwand committed:
refactor: <View> -> <>. TouchableWithoutFeedback needs children elem

digiwand committed:
fix: rm flatContainer paddingHorizontal

digiwand commented:
Created 1/3 cherry-pick PRs to support BottomSheet in 7.42.0: https://github.com/MetaMask/metamask-mobile/pull/13914
